File: Python/shortest-job-first.py
#!/usr/bin/env python3

# Author : A.F. Agarap
# Shortest Job First algorithm

jobs = []

size = int(input("How many processes will you enter? "))

for index in range(0, size):
    jobs.append(
        [
            (int(input("Enter arrival time for job #{}: ".format(index + 1)))),
            (int(input("Enter service time for job #{}: ".format(index + 1)))),
        ]
    )

jobs.sort(key=lambda x: x[1])

for index in range(0, len(jobs)):
    if index == 0:
        jobs[index].append(0)
    else:
        jobs[index].append(
            jobs[index - 1][3]
            if (jobs[index - 1][3] > jobs[index][0])
            else jobs[index][0]
        )
    jobs[index].append(jobs[index][2] + jobs[index][1])

for index in range(0, len(jobs)):
    if index == 0:
        jobs[index].append(0)
    else:
        jobs[index].append(jobs[index][2] - jobs[index][0])

for index in range(0, len(jobs)):
    jobs[index].append(jobs[index][3] - jobs[index][0])

print("\nGantt chart:\n")
for index in range(0, len(jobs)):
    if jobs[index][2] == jobs[index - 1][3]:
        for index2 in range(0, jobs[index][3] - jobs[index - 1][3]):
            print(" - ", end="")
        print(jobs[index][3], " ", end="")
    elif jobs[index][2] != jobs[index - 1][3] or i == 0:
        print(jobs[index][2], end="")
        for index2 in range(0, jobs[index][3] - jobs[index][2]):
            print(" - ", end="")
        print(jobs[index][3], " ", end="")

print("\n\nAT\tBT\tST\tFT\tWT\tTAT")
for index in range(0, len(jobs)):
    print(
        "{}\t{}\t{}\t{}\t{}\t{}".format(
            jobs[index][0],
            jobs[index][1],
            jobs[index][2],
            jobs[index][3],
            jobs[index][4],
            jobs[index][5],
        )
    )

File: C++/shortest-job-first.cpp
#include <iostream>
#include <sstream>
#include <string>

int input(std::string message);
std::string concat(std::string message, int number);

int main() {
	int temp = 0;
	int size = input("How many processes will you enter? ");

	int process[size][6];

	for (int index = 0; index < size; index++) {
		process[index][0] = input(concat("Enter arrival time for job #", index));
		process[index][1] = input(concat("Enter service time for job #", index));
		std::cout << "================================" << std::endl;
	}

	// [[0, 5], [1, 3], [2, 8], [3, 6]]
	/*
		5 3 8 6
			5 < 3 : swap 3 and 5	([0] < [1])
		3 5 8 6
			5 < 8 : as is 			([1] < [2])
		3 5 8 6	
		3 5 8 6
			8 < 6 : 				([2] < [3])
		3 5 8 6
			6 < 3 :					([3] < [0])
		3 5 8 6
			6 < 5 :					([3] < [1])
		3 5 8 6
			6 < 8 :					([3] < [2])
	*/

	for (int index = 1; index < size; index++) {
		if (process[index - 1][1] > process[index][1]) {
			temp = process[index][1];
			process[index][1] = process[index - 1][1];
			process[index - 1][1] = temp;

			temp = process[index][0];
			process[index][0] = process[index - 1][0];
			process[index - 1][0] = temp;
		}
	}

	for (int index = 0; index < size; index++) {
		std::cout << process[index][0] << " : " << process[index][1] << std::endl;
	}

	return 0;
}

// [[0, 5], [1, 3], [2, 8], [3, 6]]

int input(std::string message) {
	int number;
	std::cout << message;
	std::cin >> number;
	return number;
}

std::string concat(std::string message, int number) {
	std::stringstream ss;
	std::string s;
	ss << message << number << ": ";
	s = ss.str();
	return s;
}